### PR TITLE
codegen: start to remove the ability to call back into inference

### PR DESCRIPTION
**Pull request by vtjnash**
Continuing the development in #53219, according to the plan in https://hackmd.io/@vtjnash/codeinstances, this further separates the meaning behind CodeInfo and CodeInstance, such that CodeInstance can only be used as a call target, and cannot be used for code generation, while CodeInfo can only be used to generate code (or for reflection on what code would be generated), and cannot be used as a call target. Basically, the eventual idea is that CodeInfo will only show up now as an input (e.g. for doing inlining or codegen) and is ephemeral, while a CodeInstance is what shows up in a cache (e.g. as a callable object).

### CODE REVIEW

File: src/gf.c [hunk 1]
```diff
@@ -430,6 +430,44 @@ jl_code_instance_t *jl_type_infer(jl_method_instance_t *mi, size_t world, int fo
...
+    JL_GC_PUSHARGS(fargs, 4);
+    jl_module_t *CC = (jl_module_t*)jl_get_global(jl_core_module, jl_symbol("Compiler"));
+    if (CC != NULL && jl_is_module(CC)) {
+        fargs[0] = jl_get_global(CC, jl_symbol("NativeInterpreter"));;
+        fargs[1] = jl_box_ulong(world);
+        fargs[1] = jl_apply(fargs, 2);
+        fargs[0] = jl_get_global(CC, jl_symbol("typeinf_code"));
+        fargs[2] = (jl_value_t*)mi;
+        fargs[3] = jl_true;
+        ci = (jl_code_info_t*)jl_apply(fargs, 4);
```
Comment (vchuravy) [inline src/gf.c:456]: Why is this not calling `jl_typeinf_func`? Or `typeinf_ext_toplevel` instead of constructing the NativeInterpreter manually?

**Review by vchuravy — APPROVED** [review]
This should be fine for GPUCompiler. If you are planning to remove `lookup` from `jl_ci_cache_lookup` and thus `jl_create_native` we should talk about a strategy for GPUCompiler.

Currently GPUCompiler does: https://github.com/JuliaGPU/GPUCompiler.jl/blob/8b513be9e2230fe0dd1905b805e25fa049b24d1d/src/jlgen.jl#L603-L617

I did run test locally for GPUCompiler and all passed.

cc: @maleadt

File: src/gf.c (hunk 1 above)
Comment (vtjnash) [reply src/gf.c:456]: Those do not return source code, while the point of this function is to implement a basic version of `code_typed`, for debugging in gdb when things are looking broken

**Comment by vtjnash** [comment]
> we should talk about a strategy for GPUCompiler.

The plan is to require a CodeInstance in that IR position instead of a MethodInstance, so that the lookup is no longer necessary

**Pull request merged**